### ISSUE TITLE
OG compatibility - 2.x

### ISSUE THREAD

**Issue by paean99** [RESTful-Drupal/restful#877]
Thanks for this module.

I have tested RestFul 7.x-1.6 with Organic Group and everything seems to be ok, Unfortunately i have being having difficulties with 7.x-2.7.

When i enable 'og_access' and i make a group private, even with the admin account and being also the administrator member, the private groups and all of their content becomes inaccessible in the api.
Tested with a vanilla drupal 7.43 install and Entity patched '[2086225-entity-access-check-18.patch](https://www.drupal.org/files/issues/2086225-entity-access-check-18.patch)' , I used 'article resource' of restful_example in the tests.

Must i do something different from 1.x or is it not compatible at the moment?
From browsing the code, the two versions of RestFul are using the same method (entity_access()) to validate the access., although they differ greatly in getAccount().

**Comment by fpalluel** [comment]
Hello,
First, I built a working site with Og2.8 and restful2.7, so yes, they are compatible ;-)

What do you mean by "inaccessible" ? 403 error ?
Also, which authentication method are you using ?
Are you issuing your requests from the same host ? Beware of CORS, OPTIONS requests can cause problems...

Restful is not a "plug and play" module, but once you resolved the initial problems, it's the most flexible and powerful restful solution on Drupal, IMHO :-)

**Comment by paean99** [comment]
Here is my test:
(Drupal 7.43, Organic groups 7.x-2.7 [not the latest version, but shouldn't be relevant?], RESTful 7.x-2.7 and its RESTful example sub-module to expose the article content type in the api and thus remove the possibility of some coding mistake on my part).
- I created a new content type called GroupType and at admin/config/group/fields given him two fields group_group and group_access (from  Organic groups access control sub-module).
- I also used the Article content type as the Group content with two fields og_group_ref and group_content_access.
- Then i created a Group01 node (GroupType with 'Group visibility: Private') and a Article01 with 'Groups Audience: Group01' and 'Group content visibility: private'.
- I created a new user01 to drupal and added it as a member of Group01.

The result is that i have a group and its contents that can only be accessed to the members of the group. Using the normal drupal (access/url/routing/template) way, everything seems ok and going by the OG permissions.

> What do you mean by "inaccessible" ? 403 error ?

Unfortunately, going to api/v2.1/articles, i can only see the articles that i have created without selection of a group (not pertaining to a group) or those that are included in a group with 'Group visibility: Public'. And that with the admin or user01 account or any other. The Private groups and content are not listed.

When i try everything above with Restful 7.x-1.6, i have no problem in this simple test. The groups and groups contents are displayed to the corrects accounts via the api url.

I also tested with other content types (groups and group audiences) and exposing them to the api:, same result: works in Restful 7.x-1.6 and do not with Restful 7.x-2.7.

> Also, which authentication method are you using ?

I used the browser directly, so the authentication should be with cookies. I could test with other authentication types with a REST client... I also haven't tried other CRUD operations at this time.

> First, I built a working site with Og2.8 and restful2.7, so yes, they are compatible ;-)

It makes me happy to ear it, but it only suggest that i am doing something wrong...
@Moxide have you used 'Organic groups access control sub-module' with private groups and contents on your site?
Thanks for the feedback.

edit: updated OG to 7.x-2.8 (seeing so many commits...), but no changes :(

**Comment by fpalluel** [comment]
> have you used 'Organic groups access control sub-module' with private groups and contents on your site?

Yes all private groups

First, get a Rest client, it's priceless ! For instance, I use Http requester for Firefox.
When you're using 'api/v2.1/articles', does it refer to a custom resource file you made, or the example resource file 'Articles__2_1.php' located in the 'restful_example' folder ?

In the first case, could you post your resource file ?
In the second case, could you try 'api/1.5/articles' ?

**Comment by paean99** [comment]
@Moxide Thank you very much for your last comment, in particular the suggestion about trying version 1.5 of articles.

I was able to discover the error in my approach.
Versions of Articles resources (in Restful 2.7 examples): 1.0, 1.4, 1.5, 1.6 and 1.7 gave results ‘consistent’ with private OGs, but not versions 1.1, 2.0 and 2.1.
After that I realized that I experimented the earlier article versions, but it was just with Restful 1.6. Instead in Restful 2.7, I just used Articles v2.0 and 2.1 :(

All the ‘incompatibles’ versions of Articles resource were missing ‘authenticationTypes = TRUE’. They just had ‘authenticationOptional = TRUE’ and I was blindly using them as my template in testing.
After adding ‘authenticationTypes = TRUE’, everything at this time (Read in CRUD) seems to be working ok.

> First, get a Rest client, it's priceless ! For instance, I use Http requester for Firefox.

Yes, I usually use Chrome so I have an extension named ‘Advanced REST client’ and I am pleased with it. It is very useful for testing others authentication methods (like the basic one where I don’t even have to do the base64 encode) and other requests.
I just did some quick tests and they where the same independently of the authentication method.

I don’t mind using the normal browsing method for simple tests, because I also use a Chrome extension named ‘JSONView’ that gives me a very nice print and navigation of the json tree. But I will probably beggin to use the Rest client more intensively.
